[Static Runtime] Add pass to eliminate __getitem__/DictConstruct calls

mikeiovine opened the pull request:
Summary:
Introduce a new pass to eliminate calls to `prim::DictConstruct/aten::__getitem__`. Given a graph like this:
```
%2 : Dict = prim::DictConstruct(%key, %value)
%3 : Tensor = aten::__getitem__(%2, %key)
%4 : Tensor = op(%3)
```
This pass produces a graph like this (after dead code elimination):
```
%4 : Tensor = op(%value)
```

This optimization is applied in the static runtime.

Test Plan: New test: `buck test //caffe2/test/cpp/jit:jit`

Differential Revision: D29995087

eellison reviewed (commented):
Can you add this as a pass that is invoked in PeepholeOptimize ? It would be great to have new passes and logic added to the general mechanisms which Static Runtime benefits from, as do other consumers like TorchScript, ONNX, Mobile, etc.

eellison commented on torch/csrc/jit/passes/peephole_list_idioms.cpp:
```diff
@@ -23,13 +27,36 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+    for (size_t i = 0; i < dict_creation_node->inputs().size(); i += 2) {
+      Value* key = dict_creation_node->input(i);
+      bool is_constant = toIValue(key) != c10::nullopt;
+      if (is_constant && dict.find(key) == dict.end()) {
```
we dont want to look on `Value *` to check on overlap we want to check on overlap of the IValue value `key`. i would just handle common immutable keys for now (int, str, float) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L761

mikeiovine commented:
It's OK to leave this logic in `peephole_list_idioms` right? The alternative would be to put all dict stuff in a new pass. That'd be more modular, but less efficient for my use case (would need 2 alias DBs instead of just 1)

Commit:
[Static Runtime] Add pass to eliminate __getitem__/DictConstruct calls (#62429)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62429

Introduce a new pass to eliminate calls to `prim::DictConstruct/aten::__getitem__`. Given a graph like this:
```
%2 : Dict = prim::DictConstruct(%key, %value)
%3 : Tensor = aten::__getitem__(%2, %key)
%4 : Tensor = op(%3)
```
This pass produces a graph like this (after dead code elimination):
```
%4 : Tensor = op(%value)
```

This optimization is applied in the static runtime.

Test Plan:
`buck test //caffe2/test:jit -- TestPeephole`

**local.forward performance summary**
About 3% runtime benefit. All `DictConstruct` calls optimized out, `__getitem__` calls reduced significantly (~50% of them are cut out)
P438354810

**local_request_only.forward performance summary**
About 14% runtime benefit. Again, all `DictConstruct` calls optimized out, 50% `__getitem__` calls removed.
P438359742

There is some variance with runtime measurements, so take these numbers with a grain of salt. Also note that the benefit does not exist in the shrunk model since there are no `DictConstruct` calls

Reviewed By: hlu1

Differential Revision: D29995087

fbshipit-source-id: ef5786ae59e76637a2b27e5d93b0fd1317bff95e
